ocaml/lambda/translattribute.ml

```diff
@@ -297,24 +297,35 @@ let get_local_attribute l =
   let attr = find_attribute is_local_attribute l in
   parse_local_attribute attr
 
-let get_property_attribute l p =
+let get_property_attribute l p ~fun_attr =
   let attr = find_attribute (is_property_attribute p) l in
   let res = parse_property_attribute attr p in
   (match attr, res with
    | None, Default_check -> ()
    | _, Default_check -> ()
    | None, (Check _ | Ignore_assert_all _ ) -> assert false
    | Some _, Ignore_assert_all _ -> ()
-   | Some attr, Check _ ->
+   | Some attr, Check { assume; _ } ->
      if !Clflags.zero_alloc_check && !Clflags.native_code then
        (* The warning for unchecked functions will not trigger if the check is requested
           through the [@@@zero_alloc all] top-level annotation rather than through the
           function annotation [@zero_alloc]. *)
-       Builtin_attributes.register_property attr.attr_name);
+       if assume then begin
+         let never_specialise =
+           if Config.flambda then
+              fun_attr.specialise = Never_specialise
+           else
+              (* closure drops [@specialise never] and never specialises *)
+              true
+         in
+         if not ((fun_attr.inline = Never_inline) && never_specialise) then
+          Location.prerr_warning attr.attr_name.loc
+            (Warnings.Misplaced_assume_attribute attr.attr_name.txt)
+       end
+       else
+         Builtin_attributes.register_property attr.attr_name);
```

Comment:
If I understand this logic correctly, it's important that the `specialise` and `inline` attributes have been added to the function attrs before we run this code (that is, the order that the `add_*_attribute` functions are used in `add_function_attributes` is relevant now).  Can you add a comment somewhere saying that?

Reply:
The comment looks fine (except for a typo, see below).

I agree this example is unfortunate.  Since there is an easy work-around, this is unlikely to occur in existing code, and you have plans to rework this, I think it's fine to merge and not worth investing more time now.

Reply:
Thanks!